### PR TITLE
#78 Updated pedal fault conditions

### PR DESCRIPTION
**Pull request by bharath-v-m**
## Changes

_Explanation of changes goes here_

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: Core/Src/monitor.c [hunk 1]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
 		// MAX_ADC_VAL_12B) && 	!is_timer_active(&diff_timer)) 	start_timer(&diff_timer,
 		// PEDAL_FAULT_TIME); else 	cancel_timer(&diff_timer);
 
+
+		/* Evaluate accelerator faults */
+		if (is_timer_expired(&oc_timer))
+			//todo queue fault
+			continue;
+		else if ((adc_data[ACCELPIN_1] == MAX_ADC_VAL_12B || adc_data[ACCELPIN_2] == MAX_ADC_VAL_12B) &&
```
Comment (nwdepatie) [inline Core/Src/monitor.c:155]: nit: can we change `MAX_ADC_VAL_12B` to `MAX_ADC_VAL_12b`? The capital B is usually reserved for bytes, while lowercase is used for bits

File: Core/Src/monitor.c [hunk 2]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
 		// MAX_ADC_VAL_12B) && 	!is_timer_active(&diff_timer)) 	start_timer(&diff_timer,
 		// PEDAL_FAULT_TIME); else 	cancel_timer(&diff_timer);
 
+
+		/* Evaluate accelerator faults */
```
Comment (nwdepatie) [inline Core/Src/monitor.c:151]: Can you add in more comments in this section to explain what is actually happening? Feels like a very large chunk of logic that is a bit hard to decipher.

File: Core/Src/monitor.c [hunk 3]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
...
+		if (is_timer_expired(&oc_timer))
+			//todo queue fault
```
Comment (nwdepatie) [inline Core/Src/monitor.c:153]: Can we move all of the fault evaluation for pedals to a helper function? This is a fairly large piece of code that can be easily moved to a `static void eval_pedal_fault()` function that might help with readability of the task

File: Core/Src/monitor.c [hunk 4]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
...
+			continue;
+		else if ((adc_data[ACCELPIN_1] == MAX_ADC_VAL_12B || adc_data[ACCELPIN_2] == MAX_ADC_VAL_12B) &&
+			!is_timer_active(&oc_timer))
+			start_timer(&oc_timer, PEDAL_FAULT_TIME);
+		else
+			cancel_timer(&oc_timer);
+		
+		if (is_timer_expired(&sc_timer))
+			//todo queue fault
```
Comment (nwdepatie) [inline Core/Src/monitor.c:162]: Doesn't seem like you are queuing faults yet. I think all you need to do is just call `queue_fault(&fault_data);`

File: Core/Src/monitor.c (hunk 3 above)
Comment (nwdepatie) [reply Core/Src/monitor.c:153]: Also seems like we use the same logic for brakes and pedals, we could use the same function for both and just pass in the different values (and maybe thresholds)

File: Core/Src/monitor.c [hunk 5]
```diff
@@ -97,6 +97,14 @@ void vPedalsMonitor(void* pv_params)
 	// nertimer_t diff_timer;
 	// nertimer_t sc_timer;
 	// nertimer_t oc_timer;
+
+	/*diff_timer: timer for difference between pedal sensor values*/
+	nertimer_t diff_timer;
+	/*sc: short circuit tiimer*/
+	nertimer_t sc_timer;
+	/*oc: open circuit timer*/
+	nertimer_t oc_timer;
```
Comment (nwdepatie) [inline Core/Src/monitor.c:106]: Seems like we use the same timer for brake and pedal faults, this will pose a problem since they are both referencing the same debounce

File: Core/Src/monitor.c [hunk 6]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
 		// MAX_ADC_VAL_12B) && 	!is_timer_active(&diff_timer)) 	start_timer(&diff_timer,
 		// PEDAL_FAULT_TIME); else 	cancel_timer(&diff_timer);
 
+
+		/* Evaluate accelerator faults */
+		if (is_timer_expired(&oc_timer))
+			//todo queue fault
+			continue;
```
Comment (nwdepatie) [inline Core/Src/monitor.c:154]: I don't think you want continue here, it will jump to the next loop of the entire task's for loop so it won't actually push that data to torque. Definitely a discussion of "do we want to actually push potentially bad data to torque" but if we low pass filter the torque and pedal values it should be fine, as long as the filter is longer than the debounce time for faults

File: Core/Src/monitor.c [hunk 7]
```diff
@@ -139,6 +147,71 @@ void vPedalsMonitor(void* pv_params)
...
+			!is_timer_active(&diff_timer))
+			start_timer(&diff_timer, PEDAL_FAULT_TIME);
+		else
+			cancel_timer(&diff_timer);
+
+		/*Evaluate brake faults*/
+
+		/*Open circuit*/
+
+		if(is_timer_expired(&oc_timer))
```
Comment (nwdepatie) [inline Core/Src/monitor.c:183]: I think we need to review this faulting logic. Right now the timer will never actually reset itself if we fix the fault. See my implementation of the pedal reading for last year's car: https://github.com/Northeastern-Electric-Racing/Embedded_Code/blob/main/MPU/src/pedalHW.cpp

**Review by nwdepatie — CHANGES_REQUESTED** [review]
There's some funky stuff going on, check my comments

File: Core/Src/monitor.c (hunk 1 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: Updated accordingly

File: Core/Src/monitor.c (hunk 2 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: Added necessary comments. If needed, will make it more comprehensive.

File: Core/Src/monitor.c (hunk 3 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: added a helper function. But it has extra parameters - void eval_pedal_fault(Sensor_1, Sensor_2, diff_timer, sc_timer, oc_timer). I was wondering if the interface can be updated accordingly

File: Core/Src/monitor.c (hunk 4 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: added the queuing faults.

File: Core/Src/monitor.c (hunk 5 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: i've added custom timers for the accelerator and brake sensors separately.

File: Core/Src/monitor.c (hunk 7 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: reviewed the code and made it more verbose

File: Core/Src/monitor.c (hunk 6 above)
Comment (bharath-v-m) [reply Core/Src/monitor.c]: removed continue

**Review by nwdepatie — APPROVED** [review]
A bit sketched out but we need this in ASAP, we can verify this at a later date when the hardware is more put together. I also gotta review the rules again to make sure this is 110% what we need
